customizable section snippet

jmisur opened the pull request:
Fix for https://github.com/ScaCap/spring-auto-restdocs/issues/63
- support for custom sections in the snippet and their order
- support for skipping empty sections

In this way, any additional Snippet could be added to the section template via implementing SectionSupport and including it's snippetName in SectionBuilder.snippetNames.

TODO: update docs

fbenz commented:
We have to be aware that this is a breaking change. Existing users of the section snippet like our example project have to adjust their configuration. The breaking change could be avoided by detecting the original Spring REST Docs snippets and transforming/treating them as valid snippets for the section snippet.

jmisur commented:
I removed those custom subclasses and added support for classic snippets (backwards compat).
Also update docs.

fbenz commented on spring-auto-restdocs-docs/section-snippet.adoc:
```diff
@@ -29,41 +29,84 @@ include::../../../spring-auto-restdocs-core/src/main/resources/org/springframewo
 
 |===
 
+If you want to configure custom sections and thier order, instead of using `AutoDocumentation.section()` you can use
```
typo: `thier` -> `their`

fbenz commented on spring-auto-restdocs-docs/section-snippet.adoc:
```diff
@@ -29,41 +29,84 @@ include::../../../spring-auto-restdocs-core/src/main/resources/org/springframewo
+    .build()
+----
+<1> only selected snippets will be used in the specified order
+<2> if snippets having no content should be skipped from section
```
`whether` is better than `if` here

fbenz commented on spring-auto-restdocs-docs/section-snippet.adoc:
```diff
@@ -29,41 +29,84 @@ include::../../../spring-auto-restdocs-core/src/main/resources/org/springframewo
+
+| {sections}
+| <iterable>
+| Contains all sections (Snippet classes) which were configured to be included in the section snippet. Moustache templater iterates over this and expands into separate header and link lines.
```
`were configured` -> `are configured`

jmisur commented:
moved section related stuff to separate package and SnippetRegistry to root package. This will become public API, so I also hid the registry impl a bit. 
Also updated docu. pls check

fbenz commented on spring-auto-restdocs-docs/section-snippet.adoc:
```diff
@@ -109,4 +113,13 @@ AutoDocumentation.sectionBuilder()
-<2> if snippets having no content should be skipped from section
+<2> whether snippets having no content should be skipped from section
+
+==== Custom snippet in section
```
I would call it: Custom snippet as a section

jmisur replied:
it's actually custom snippet inside a section snippet. it's confusing because section snippet contains multiple snippet "sections" :-/

fbenz replied:
"Custom snippet as a section in the section snippet" ;) But as this part of the documentation is only about the section snippet it might be clearer if we talk about adding a section (to the section snippet).

fbenz commented on spring-auto-restdocs-docs/section-snippet.adoc:
```diff
@@ -109,4 +113,13 @@ AutoDocumentation.sectionBuilder()
+
+Custom snippets are created by subclassing a link:{restdocs-package}/snippet/Snippet.java[Snippet] interface
+(or more conveniently link:{restdocs-package}/snippet/TemplatedSnippet.java[TemplatedSnippet]). If you want your custom
+ snippet to be included in section and automatically recognised header and include link (file name),
```
included in section and automatically recognised header and include link -> included as a section with an automatically recognised header and an include link

jmisur commented:
ok did a bit of compromise. we can improve later.
